http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala

```diff
@@ -0,0 +1,194 @@
+import org.broadinstitute.dsde.workbench.model.WorkbenchException
+import cats.implicits._
+
+import scala.concurrent.ExecutionContext
+
+case class ClusterNotFoundException(
+  clusterId: Long,
+  override val message: String =
+    "Could not process ClusterTransitionFinishedMessage because it was not found in the database"
+) extends LeoException
+
+class LeoPubsubMessageSubscriber[F[_]: Async: Timer: ContextShift: Logger: Concurrent](
+  subscriber: GoogleSubscriber[IO, LeoPubsubMessage],
+  clusterHelper: ClusterHelper,
+  dbRef: DbReference[IO]
+)(implicit executionContext: ExecutionContext)
+    extends LazyLogging {
+
+  def messageResponder(message: LeoPubsubMessage): IO[Unit] = {
+    val response = message match {
+      case msg @ StopUpdateMessage(_, _) =>
+        handleStopUpdateMessage(msg)
+      case msg @ ClusterTransitionFinishedMessage(_) =>
+        handleClusterTransitionFinished(msg)
+      case _ => IO.unit
+    }
+
+    //ensure we don't crash if the handler throws an exception
+    response.onError {
+      case e => IO(logger.error(s"Unable to process a message in received from pub/sub subscription in messageResponder: ${message}, errorMessage: ${e.getMessage}", e))
+    }
+  }
+
+  private def messageHandler: Pipe[IO, Event[LeoPubsubMessage], Unit] = in => {
+    in.flatMap { event =>
+      for {
+        _ <- Stream.eval(messageResponder(event.msg))
+        _ <- Stream.eval(IO.pure(event.consumer.ack()))
+      } yield ()
+    }
+  }
+
+  val process: Stream[IO, Unit] = (subscriber.messages through messageHandler).repeat
+
+  private def handleStopUpdateMessage(message: StopUpdateMessage): IO[Unit] =
+    dbRef
+      .inTransaction { clusterQuery.getClusterById(message.clusterId) }
+      .flatMap {
+        case Some(resolvedCluster)
+            if ClusterStatus.stoppableStatuses.contains(resolvedCluster.status) && !message.updatedMachineConfig.masterMachineType.isEmpty => {
+          val followupDetails = ClusterFollowupDetails(message.clusterId, ClusterStatus.Stopped)
+          logger.info(
+            s"stopping cluster ${resolvedCluster.projectNameString} in messageResponder, and saving a record for ${resolvedCluster.id}"
+          )
+          for {
+            _ <- dbRef.inTransaction(
+              followupQuery.save(followupDetails, message.updatedMachineConfig.masterMachineType)
+            )
+            _ <- clusterHelper.stopCluster(resolvedCluster)
+          } yield ()
+        }
+        case Some(resolvedCluster) =>
+          IO.raiseError(
+            new WorkbenchException(
+              s"Failed to process StopUpdateMessage for Cluster ${resolvedCluster.projectNameString}. This is likely due to a mismatch in state between the db and the message, or an improperly formatted machineConfig in the message. Cluster details: ${resolvedCluster}"
+            )
+          )
+        case None =>
+          IO.raiseError(
+            new WorkbenchException(
```

Comment:
Minor comment, but I see you use `ClusterNotFoundException` in another place for this, but `WorkbenchException` here. Probably just using `WorkbenchException` is fine.